packages/clarity-tutorials/contracts/rocket-tutorial/rocket-market.scm

```diff
@@ -0,0 +1,138 @@
+;;  copyright: (c) 2013-2019 by Blockstack PBC, a public benefit corporation.
+
+;;  This file is part of Blockstack.
+
+;;  Blockstack is free software. You may redistribute or modify
+;;  it under the terms of the GNU General Public License as published by
+;;  the Free Software Foundation, either version 3 of the License or
+;;  (at your option) any later version.
+
+;;  Blockstack is distributed in the hope that it will be useful,
+;;  but WITHOUT ANY WARRANTY, including without the implied warranty of
+;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;;  GNU General Public License for more details.
+
+;;  You should have received a copy of the GNU General Public License
+;;  along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+
+;;;; Rocket-Market
+
+;;; Storage
+(define-map rockets-info
+  ((rocket-id int)) 
+  ((owner principal)))
+(define-map rockets-count
+  ((owner principal))
+  ((count int)))
+(define-map factory-address
+  ((id int))
+  ((address principal)))
+
+;;; Internals
+
+;; Gets the amount of rockets owned by the specified address
+;; args:
+;; @account (principal) the principal of the user
+;; returns: int
+(define (balance-of (account principal))
+  (let ((balance
+      (get count 
+        (fetch-entry rockets-count (tuple (owner account))))))
+    (if (eq? balance 'null) 0 balance)))
+
+;; Check if the transaction has been sent by the factory-address
+;; returns: boolean
+(define (is-tx-from-factory)
+  (let ((address
+    (get address 
+        (fetch-entry factory-address (tuple (id 0))))))
+    (eq? tx-sender address)))
+
+;; Gets the owner of the specified rocket ID
+;; args:
+;; @rocket-id (int) the id of the rocket to identify
+;; returns: principal
+(define (owner-of (rocket-id int)) 
+  (get owner 
+    (fetch-entry rockets-info (tuple (rocket-id rocket-id)))))
+
+;; Register a rocket
+;; args:
+;; @new-owner (principal) the principal of the new owner
+;; @rocket-id (int) the id of the rocket to transfer
+;; returns: boolean
+(define (register-rocket! (new-owner principal) (rocket-id int))
+  (let ((current-balance (balance-of new-owner)))
+    (begin
+      (set-entry! rockets-info 
+        (tuple (rocket-id rocket-id))
+        (tuple (owner new-owner))) 
+      (set-entry! rockets-count 
+        (tuple (owner new-owner))
+        (tuple (count (+ 1 current-balance)))) 
+      'true)))
+
+;; Release a rocket
+;; args:
+;; @owner (principal) the principal of the current owner
+;; @rocket-id (int) the id of the rocket to release
+;; returns: boolean
+(define (release-rocket! (owner principal) (rocket-id int))
+  (let ((current-balance (balance-of owner)))
+    (begin
+      (set-entry! rockets-count 
+        (tuple (owner owner))
+        (tuple (count (- current-balance 1)))) 
```

Comment:
Underflow check?

Reply:
Didn't we settled on the idea of having underflows (int/uint) throwing exceptions in order to avoid having developers doing these checks?

Reply:
I think we discussed the possibility of having an unsigned integer type in addition to a signed integer type, but I don't think this has been decided yet (cc @kantai).

Reply:
Regardless, since this code is something that developers are going to learn from, it's best to code defensively in order to help new developers cultivate good habits ;)

Reply:
As a developer, if I'm reading in a documentation that `underflows / overflows` are throwing, and then going over a sample code that tests for preventing `underflows / overflows`, I think I would be very confused.

Reply:
I agree that we should allow underflows and overflows to throw exceptions here -- I think that it makes the contract code itself more clear, and is also instructive of what happens in those events.

However, I think, here, we should check that current-balance is > 0 before allowing the function to decrement a rocket (though, that check _is_ happening in the only place `release-rocket!` can be called).

Reply:
Thank you @kantai!

After re-reading the code, I think I don't like the idea of splitting the transfer in 2 methods, mostly because `release-token` does not make sense in this context: we are decrementing the balance, without updating the `rocket-info` map.
Update available here - https://github.com/blockstack/clarity-js-sdk/pull/4/commits/9ee0169adbe1e6a848ee6a5524f6b73c9af2d9b1